tests/README.rst

```diff
@@ -0,0 +1,124 @@
+tests might be useful.  Others might just be interested in how a testing
+library tests itself!
+
+
+The very short version
+======================
+
+- To improve code coverage (eg because ``make check-coverage`` / the Travis
+  build is failing), go to ``cover/``
+- For longer / system / integration tests, look in ``nocover/``
+- For tests that require an optional dependency, look in the directory
+  named for that dependency.
+
+.. note::
+    If you get stuck, just ask a maintainer to help out by mentioning
+    ``@HypothesisWorks/hypothesis-python-contributors`` on GitHub.
+    We'd love to help - and also get feedback on how this document could
+    be better!
+
+
+Some scenarios
+==============
+
+**I'm adding or changing a strategy**
+    Check for a file specific to that strategy (eg ``test_uuids.py`` for
+    the ``uuids()`` strategy).  Write tests for all invalid argument handling
+    in ``test_direct_strategies.py``.  Strategies with optional dependencies
+    should go in ``hypothesis.extras``, and the tests in their own module
+    (ie not in ``cover``).  When you think you might be done, push and let
+    Travis point out any failing tests or non-covered code!
+
+**I've made some internal changes**
+    That's not very specific - you should probably refer to the test-finding
+    tips in the next section.  Remember that ``tests/cover`` is reasonably
+    quick unit-test style tests - you should consider writing more intensive
+    integration tests too, but put them in ``tests/nocover`` with the others.
+
+
+Finding particular tests
+========================
+
+With the sheer size and variety in this directory finding a specific thing
+can be tricky.  Tips:
+
+- Check for filenames that are relevant to your contribution.
+- Use ``git grep`` to search for keywords, e.g. the name of a strategy you've changed.
+- Deliberately break something related to your code, and see which tests fail.
+- Ask a maintainer!  Sometimes the structure is just arbitrary, and other tactics
+  don't work - but as I keep saying, we *want* to help!
+
+
+About each group of tests
+=========================
+
+Still here?  Here's a note on what to expect in each directory.
+
+``common/``
+    Useful shared testing code, including test setup and a few helper
+    functions in ``utils.py``.  Also read up on
+    `pytest <https://docs.pytest.org/en/latest/contents.html>`_
+    features such as ``mark.parametrize``, ``mark.skipif``, and ``raises``
+    for other functions that are often useful when writing tests.
+
+``cover/``
+    The home of enough tests to get 100% branch coverage - you'll probably
+    spend most of your time here.  Ideally this would be as quick as possible
+    while remaining comprehensive.  More details inside.
+
+    This directory alone has around two-thirds of the tests for Hypothesis
+    (~8k of ~12k lines of code).  If you're adding or fixing tests, chances
```

Comment:
I think we should actually default to adding new tests to nocover, and only add new tests to cover when they are required to cover some uncovered part of the code (we/I haven't historically been good about this, but I think it would make life better).